libraries/botbuilder-lg/src/activityFactory.ts

```diff
@@ -0,0 +1,361 @@
+        return buttons;
+    }
+
+    private static getCardAction(cardType: string, cardActionJObj: any): CardAction {
+        const type: string = this.getStructureType(cardActionJObj);
+        let cardAction: any = {};
+
+        if (cardType === CardFactory.contentTypes.signinCard || cardType === CardFactory.contentTypes.oauthCard) {
+            cardAction.type = ActionTypes.Signin;
+        } else {
+            cardAction.type = ActionTypes.ImBack;
+        }
+
+        if(type !== 'cardaction') {
+            return undefined;
+        }
+
+        for (const key of Object.keys(cardActionJObj)) {
+            const property: string = key.trim();
+            const value: any = cardActionJObj[key];
+
+            switch (property.toLowerCase()) {
+                case 'type':
+                    cardAction.type = value.toString();
+                    break;
+                case 'title':
+                    cardAction.title = value.toString();
+                    break;
+                case 'value':
+                    cardAction.value = value.toString();
+                    break;
+                case 'displaytext':
+                    cardAction.displayText = value.toString();
+                    break;
+                case 'text':
+                    cardAction.text = value.toString();
+                    break;
+                case 'image':
+                    cardAction.image = value.toString();
+                    break;
+                default:
+                    console.log(`Skipping unknown activity property ${ property }`);
+                    break;
+            }
+        }
+
+        return cardAction as CardAction;
+    }
+
+    private static getStructureType(jObj: any): string {
+        let result = '';
+
+        if (jObj !== undefined) {
+            if ('$type' in jObj) {
+                result = jObj['$type'].toString();
+            } else if ('type' in jObj) {
+                result = jObj['type'].toString();
+            }
+        }
+
+        return result.trim().toLowerCase();
+    }
+
+    private static getAttachments(value: any): Attachment[] {
+        let attachments = [];
+        let attachmentsJsonList: any[] = this.normalizedToList(value);
+
+        for (const attachmentsJson of attachmentsJsonList) {
+            if (typeof attachmentsJson === 'object') {
+                const attachment = this.getAttachment(attachmentsJson);
+                if (attachment !== undefined) {
+                    attachments.push(attachment);
+                }
+            }
+        }
+
+        return attachments;
+    }
+
+    private static getAttachment(lgJObj: any): Attachment {
+        let attachment: any = {};
+        const type: string = this.getStructureType(lgJObj);
+        if (ActivityFactory.GenericCardTypeMapping.has(type)) {
+            attachment = this.getCardAttachment(ActivityFactory.GenericCardTypeMapping.get(type), lgJObj);
+        } else if(type === 'adaptivecard') {
```

Comment:
>type [](start = 18, length = 4)

is this always lower-cased? I didn't see anything in getStructureType() that does a lower case but maybe it's always lower case coming back from LG.

Reply:
yes, there is a 'toLowerCase()' in getStructuredType()